samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java

```diff
@@ -16,85 +16,378 @@
-    return new SessionWindow<>(sessionKeyFunction, (m, c) -> {
-        c.add(m);
-        return c;
-      }
-    );
+
+  public static <M extends MessageEnvelope, K> WindowFunction<M, K, WindowKey<K>, Collection<M>, WindowOutput<WindowKey<K>, Collection<M>>> keyedSessionWindow(Function<M, K> keyFn, Duration sessionGap) {
+
+    BiFunction<M, Collection<M>, Collection<M>> aggregator = (m, c) -> {
+      c.add(m);
+      return c;
+    };
+    return keyedSessionWindow(keyFn, sessionGap, aggregator);
   }
 
+
   /**
-   * Static API method to create a {@link SessionWindow} in which the output value is a collection of {@code SI} from the input {@link MessageEnvelope}s
-   *
-   * @param sessionKeyFunction  function to calculate session window key
-   * @param sessionInfoExtractor  function to retrieve session info of type {@code SI} from the input {@link MessageEnvelope} of type {@code M}
-   * @param <M>  type of the input {@link MessageEnvelope}
-   * @param <WK>  type of the session window key
-   * @param <SI>  type of the session information retrieved from each input {@link MessageEnvelope} of type {@code M}
-   * @return  the {@link Window} function for the session
+   * Returns a {@link WindowFunction} that windows all values into a single global window. This window does not have a
+   * default trigger. The triggering behavior must be specified by setting early triggers using the APIs
+   * of {@link org.apache.samza.operators.triggers.Triggers}.
+
+   * <p>The below example windows computes the maximum value over a count based window. The window emits results when
+   * there are either 50 messages in the window or when 10 seconds have passed since the first message.
+   *
+   * <pre> {@code
+   * MessageStream<Long> stream = ...;
+   * BiFunction<Long, Long, Long> maxAggregator = (m, c)-> Math.max(m, c);
+   * MessageStream<WindowOutput<WindowKey, Long>> windowed = stream.window(Windows.globalWindow(maxAggregator)
+   * .setEarlyTriggers(Triggers.repeat(Triggers.any(Triggers.count(50), Triggers.timeSinceFirstMessage(Duration.seconds(10))))))
+   * }
+   * </pre>
+   *
+   * @param aggregateFunction the function to aggregate window results
+   * @param <M> the type of {@link MessageEnvelope}
+   * @param <WV> type of the {@link WindowOutput} output value
+   * @return the created {@link WindowFunction} function.
    */
-  public static <M extends MessageEnvelope, WK, SI> Window<M, WK, Collection<SI>, WindowOutput<WK, Collection<SI>>> intoSessions(Function<M, WK> sessionKeyFunction,
-      Function<M, SI> sessionInfoExtractor) {
-    return new SessionWindow<>(sessionKeyFunction, (m, c) -> {
-        c.add(sessionInfoExtractor.apply(m));
-        return c;
-      }
-    );
+  public static <M extends MessageEnvelope, WV> WindowFunction<M, Void, WindowKey<Void>, WV, WindowOutput<WindowKey<Void>, WV>> globalWindow(BiFunction<M, WV, WV> aggregateFunction) {
```

Comment:
Question: is this defining a window w/o start and end? I don't understand the use case here. Isn't this practically defining an infinite window which is == the whole stream?